### ISSUE TITLE
Chaos monkey module

### ISSUE THREAD

**Issue by daviddenton** [http4k/http4k#118]
As touched on in the "Server as a Function. In Kotlin" presentation, it would be awesome to have a separate module which introduced bad behaviour to http4k applications.

This could be easily modelled by a set of filters which generated the chaos behaviours in http4k apps.

Inspired by this: https://codecentric.github.io/chaos-monkey-spring-boot/

**Comment by IgorPerikov** [comment]
Hi @daviddenton! I'd like to work on this one. I've started looking for http4k features, and I can't find whether it's possible to inject filter conditionally without touching application code or not? It doesn't feel right, if one will explicitly put his application handler in chain with latency injection filter and enable/disable it by passing java/env var.

**Comment by daviddenton** [comment]
Thanks for the offer of help @IgorPerikov! You're correct - there currently isn't a way to conditionally inject a filter into the stack. Then again, part of this feature idea was to determine exactly what and how we can do to inject "bad behaviour" into an application.

Initial thoughts were that this was going to evolve through several iterations:

1. Initially, identify the set of the behaviours we're like to model (latency/random failure etc)
2. Implement a set of Filters to inject these behaviours in a purely random fashion.
3. Create some kind of more detailed model/DSL for configuring the randomness - not sure if this might vary based upon the specific behaviour. For example, percentage occurance, initial delay, failure interval or activation predicate.
4. Potentially investigate doing injection of this behvaiour at runtime instead of compile time - we're not a big proponent of  classloading magic, so we'd like to avoid diving right into the deep end on this right away.

Th nice thing about the HttpHandler and Filter model is that it works at several layers of abstraction (endpoint/module/entire app) so instrumenting a running version of your http4k application can be done without implementing step 4 above.

**Comment by IgorPerikov** [comment]
@daviddenton Before I wrote my message, I had already written draft of latency injector (with exception throwing filter in mind)
And here's brief description of what I've coded - latency filter implementation which can be constructed in a pretty old-style fashion with companion method which gather some env variables like that:
`CHAOS_LATENCY_INJECTION_ENABLED`
`CHAOS_LATENCY_INJECTION_MIN_DELAY`
`CHAOS_LATENCY_INJECTION_MAX_DELAY`
`CHAOS_LATENCY_INJECTION_RATE`
or explicitly through passing these to constructor and then you simply apply it to route/app etc

So, I came up with the idea that you add your filter through the code and then enable/disable it through envvar. The reason I chose env variables (e.g. not java system vars) is simple: because of the extensive docker and virtualization usage it's much easier to pass necessary env variables through deployment configuration, instead of rewriting docker run command, pushing new image, etc

I'll show this draft tonight through pr, so we can discuss more specifically

**Comment by IgorPerikov** [comment]
@daviddenton Anything else I could help here with?

**Comment by daviddenton** [comment]
@IgorPerikov Not for the moment. I'm going to take a look, do any harmonisation, and then have a play with the Policy model now that I've got a bit of time - maybe actually using the module in anger to see what drops out of there. Once that's done we can look to releasing a first cut.

**Comment by daviddenton** [comment]
Hey. I've had a bit of a play and wanted you to see the changes first. :)

Changes are now on master - check out the tests for usage examples and let me know what you think about the API. I'm not too sure about the Filter creation.

Also keen to know if there's anything else you think we're missing - I think we definitely still need some kind of "activation" mechanism - but I'm not sure if this could be done using JAR replacement/reflective instantiation on startup.

thanks again for your work! :) :)

**Comment by IgorPerikov** [comment]
Hiya, I'll take a look and play with it in a few days. Thanks for letting me know and asking about my thoughts, that means a lot to me.

**Comment by IgorPerikov** [comment]
`ChaosControls` is for sure my favourite thing in the new API. And since it allows to activate/deactivate chaos numerous times I feel like `ChaosStage` lacks support for deactivation, we can postpone start of chaos behaviour with `Wait.until`, but we can't stop it after that (or am I missing something?).

And the main thing I am still wondering about - how one would use it as part of his daily routine, which refers third paragraph of your last message. Right now the chaos mechanisms seems imperative - you have to add all the stuff explicitly to your codebase, though it doesn't solve any of your business logic and it would be nice to be able to control it completely declaratively (twelve factor stuff again). If one want to apply chaos on real traffic, she can deploy it to canary env and not deploy it to usual production cluster, but it's two codebase versions (I bet one day one would ruin branches, or deploy wrong version to the wrong env) + some extra work building security layer on top of chaos control router (and probably implementing ACL?). 
Maybe it's possible to intercept at the `toServer()` call site, since it takes HttpHandler, and pass the parsed configuration there?

Apart from this - I guess you squeeze out the best from the imperative way of doing this.

**Comment by daviddenton** [comment]
You're right - the main thing is how we would go about using it in a production environment - definitely needs more thinking about. My initial usage of this is going to be building it into fake 3rd party services (where we can hard code it) which we use to test how our services react to outages. I shall have a further think and play...

Oh, and you were missing something with the ChaosStage - calling inject() gives you another stage which you can then call until() on. Check out the Implementation of the chaoscontrols - it models on/off behaviour using a repeating loop and a switch.

**Comment by daviddenton** [comment]
Hi @IgorPerikov - after a long period of radio silence, I've made some more changes to the module which I hope you might like. :)

The main change now is that adding `ChaosControls` to an app now adds an `OpenAPI` module, in which you can activate, deactivate, and now dynamically POST the chaos to be injected by sending some JSON to `/chaos/activate/new`. Example of the JSON is:

```[
  {
    "type": "repeat",
    "stages": [
      {
        "until": "wait",
        "trigger": {
          "type": "delay",
          "period": "PT60s"
        }
      },
      {
        "type": "only",
        "trigger": {
          "type": "deadline",
          "endTime": "1970-01-01T00:00:00Z"
        },
        "behaviour": {
          "type": "status",
          "status": 404
        }
      },
      {
        "type": "block",
        "trigger": {
          "type": "request",
          "method": "get",
          "path": ".*bob",
          "queries": {
            "query": ".*query"
          },
          "headers": {
            "header": ".*header"
          },
          "body": ".*body"
        },
        "until": {
          "type": "countdown",
          "count": "5"
        }
      }
    ]
  }
]```

I've got to write up the docs to introduce the feature.

**Comment by daviddenton** [comment]
You make a good point! I'll make that change to the docs now.

ps. Although I said the module was scheduled for 3.36.0, we actually released it in 3.35.1 so it's there if you want to have a play - thought it deserved it's own release number though. 36 is coming up soonish so I'll keep your credit in that version

Thanks for your help and feedback for this one :)
